pkg/reaper/cordon/aznat/aznat.go

```diff
@@ -0,0 +1,267 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aznat
+
+import (
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/keikoproj/governor/pkg/reaper/common"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.New()
+
+func Run(args *Args) error {
+	log.SetFormatter(&logrus.TextFormatter{
+		FullTimestamp: true,
+	})
+
+	if err := args.validate(); err != nil {
+		return errors.Wrap(err, "failed to validate arguments")
+	}
+
+	var config aws.Config
+	config.Region = &args.Region
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+		Config:            config,
+	}))
+
+	ec2Client := ec2.New(sess)
+
+	ctx := CordonContext{
+		EC2: ec2Client,
+	}
+
+	if err := ctx.discover(args.TargetVPCID); err != nil {
+		return errors.Wrap(err, "failed to discover cloud")
+	}
+
+	if err := ctx.execute(args.TargetAvailabilityZoneIDs, args.Restore, args.DryRun); err != nil {
+		return errors.Wrap(err, "failed to execute operation")
+	}
+
+	return nil
+}
+
+func (c *CordonContext) execute(targetZones []string, restore, dryRun bool) error {
+	var (
+		routes     []Route
+		zoneMap    = c.zoneGateways()
+		gwStatuses GatewayStatuses
+	)
+
+	if restore {
+		log.Infof("running route restore operation on zones: %+v, dry-run: %t", targetZones, dryRun)
+		routes = c.uncordonedRoutes(targetZones)
+	} else {
+		gwStatuses = c.gatewayStatuses(targetZones)
+		if len(gwStatuses.Active) == 0 {
+			log.Errorf("could not find active nat-gateways excluding %v, cannot uncordon all zones", targetZones)
+			return errors.Errorf("no active nat-gateways found")
+		}
+
+		log.Infof("running route cordon operation on zones: %+v, dry-run: %t", targetZones, dryRun)
+		routes = c.cordonedRoutes(gwStatuses)
+	}
+
+	if len(routes) == 0 {
+		log.Info("could not find routes to cordon/uncordon")
+	}
+
+	// replace gateway for cordoned routes
+	for _, r := range routes {
+		if restore {
+			if gws, ok := zoneMap[r.ZoneID]; ok {
+				r.NewGateway = gws[0]
```

Comment:
What if there are multiple NATs in the same AZ? Could the original NAT be remembered by adding a tag? If a different NAT is restored than what was used originally then perhaps dependencies that whitelist a specific NAT IP will be broken? Perhaps this can be done as a followup feature?

Reply:
Technically there is no difference between 1 NAT and another as long as they are in the same zone and are both healthy, hence we just chose one of existing NATs.
We can add tagging, but I am not sure there are use cases where that matters